addons/binding/org.openhab.binding.smappee/src/main/java/org/openhab/binding/smappee/internal/SmappeeService.java

```diff
@@ -0,0 +1,461 @@
+
+    private String serviceLocationId;
+
+    private boolean initialized;
+
+    private final Gson gson;
+
+    private String accessToken;
+    private DateTime accessTokenValidity;
+    private String refreshToken;
+
+    public SmappeeService(String clientId, String clientSecret, String username, String password,
+            String serviceLocationName, long pollTime) {
+        this.serviceLocationName = serviceLocationName;
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.username = username;
+        this.password = password;
+
+        this.retry = 1;
+        this.pollTime = pollTime;
+
+        this.gson = new Gson();
+
+        this.initialized = false;
+    }
+
+    public void startAutomaticRefresh(ScheduledExecutorService scheduledExecutorService,
+            final ReadingsUpdate readingsUpdate) {
+        Runnable runnable = () -> {
+            SmappeeDeviceReading readings = getDeviceReadings();
+            if (readings.consumptions.length > 0) {
+                readingsUpdate.newState(readings);
+            }
+        };
+
+        scheduledJob = scheduledExecutorService.scheduleAtFixedRate(runnable, 0, this.pollTime, TimeUnit.MILLISECONDS);
+    }
+
+    public void stopAutomaticRefresh() {
+        scheduledJob.cancel(true);
+    }
+
+    /**
+     * Gets the smappee energy consumption readings.
+     *
+     * @return the device readings
+     */
+    public SmappeeDeviceReading getDeviceReadings() {
+        int currentTry = 0;
+        do {
+            try {
+                DateTime nowUtc = DateTime.now(DateTimeZone.UTC).minusMinutes(1);
+                DateTime nowUtcMinus20Min = DateTime.now(DateTimeZone.UTC).minusMinutes(20);
+
+                String nowUtcMillis = String.valueOf(nowUtc.getMillis());
+                String nowUtcMinus20MinMillis = String.valueOf(nowUtcMinus20Min.getMillis());
+
+                // sample API method to call :
+                // https://app1pub.smappee.net/dev/v1/servicelocation/123/consumption?aggregation=1&from=1388534400000&to=1391212800000
+
+                String responseReadings = GetData("/dev/v1/servicelocation/" + this.serviceLocationId
+                        + "/consumption?aggregation=1&from=" + nowUtcMinus20MinMillis + "&to=" + nowUtcMillis);
+
+                SmappeeDeviceReading readings = gson.fromJson(responseReadings, SmappeeDeviceReading.class);
+
+                // Sum of 5 minutes in Wh so average power is 60/5 x
+
+                for (SmappeeDeviceReadingConsumption consumption : readings.consumptions) {
+                    consumption.consumption = consumption.consumption * 12;
+                    consumption.solar = consumption.solar * 12;
+                }
+
+                logger.debug("smappee'{}' read", this.serviceLocationId);
+
+                return readings;
+
+            } catch (Exception se) {
+                logger.error("failed to read smappee '{}'", se.getMessage());
```

Comment:
Don't catch Exception, but only specific errors and don't log to error. And always use the exception as last argument (not the message or tostring) and for the last argument, the exception, don't use {} in the string.

Reply:
ok, refactored